packages/react-aria-components/src/HiddenDateInput.tsx

```diff
@@ -0,0 +1,140 @@
+}
+
+export interface HiddenDateAria {
+  /** Props for the container element. */
+  containerProps: React.HTMLAttributes<HTMLDivElement>,
+  /** Props for the hidden input element. */
+  inputProps: React.InputHTMLAttributes<HTMLInputElement>
+}
+
+export function useHiddenDateInput(props: HiddenDateInputProps, state) {
+  let {
+    autoComplete,
+    isDisabled,
+    name
+  } = props;
+  let [dateValue, setDateValue] = React.useState('');
+  let {visuallyHiddenProps} = useVisuallyHidden();
+
+  let inputStep = 60;
+  if (state.granularity === 'second') {
+    inputStep = 1;
+  } else if (state.granularity === 'hour') {
+    inputStep = 3600; 
+  }
+
+  useEffect(() => {
+    if (state.value == null) {
+      setDateValue('');
+    } else {
+      setDateValue(state.value);
+    }
+  }, [state.value]);
+  
+  let inputType = state.granularity === 'day' ? 'date' : 'datetime-local';
+
+  let dateSegments = ['day', 'month', 'year'];
+  let timeSegments = ['hour', 'minute', 'second'];
+  let granularityMap = {'hour': 1, 'minute': 2, 'second': 3};
+
+  // Depending on the granularity, we only want to validate certain time segments
+  let end = 0;
+  if (timeSegments.includes(state.granularity)) {
+    end = granularityMap[state.granularity];
+    timeSegments = timeSegments.slice(0, end);
+  }
+
+  return {
+    containerProps: {
+      ...visuallyHiddenProps,
+      'aria-hidden': true,
+      // @ts-ignore
+      ['data-react-aria-prevent-focus']: true,
+      // @ts-ignore
+      ['data-a11y-ignore']: 'aria-hidden-focus'
+    },
+    inputProps: {
+      tabIndex: -1,
+      autoComplete,
+      disabled: isDisabled,
+      type: inputType,
+      // We set the form prop to an empty string to prevent the hidden date input's value from being submitted
+      form: '',
+      name,
+      step: inputStep,
+      value: dateValue,
+      onChange: (e) => {
+        let targetString = e.target.value.toString();
+        if (targetString) {
+          let targetValue: CalendarDateTime | CalendarDate = parseDateTime(targetString);
+          if (state.granularity === 'day') {
+            targetValue = parseDate(targetString);
+          }
+          setDateValue(targetString);
+          // We check to to see if setSegment exists in the state since it only exists in DateFieldState and not DatePickerState.
+          // The setValue method has different behavior depending on if it's coming from DateFieldState or DatePickerState.
+          // In DateFieldState, setValue firsts checks to make sure that each segment is filled before committing the newValue 
+          // which is why in the code below we first set each segment to validate it before committing the new value. 
+          // However, in DatePickerState, since we have to be able to commit values from the Calendar popover, we are able to set a new value
+          // when the field itself is empty. 
+          if (typeof state.setSegment === 'function') {
```

Comment:
a little unsure if this is the direction we want to go. i left some comments in the code to explain what's going on but essentially the issue is that depending if the state is DateFieldState or DatePickerState, the setValue method behaves a bit differently.

i've tried validating each segment inside setValue in useDateFieldState but the problem is that we aren't able to differentiate between a newValue that is the placeholder date versus something external (like in this case). as a result, we end up committing the placeholder value as an actual value that we want to display. in any case, it causes some really strange behavior when trying to validate segments inside setValue so it worked better to do it here inside HiddenDateInput